refactor(es/parser): Split parser into also-lex/parse-only

bvanjoi opened the pull request:
This constitutes a substantial code revision. We should first evaluate whether it yields measurable performance enhancements.

bvanjoi committed:
chore(es/parser): custom lexer

bvanjoi committed:
chore: use parser/Lexer rather than lexer/Lexer

bvanjoi committed:
chore(es/parser): split token value and token kind

bvanjoi committed:
chore(es/parser): dont clone token value

bvanjoi committed:
faster bin/assign op check

bvanjoi committed:
faster is_keyword/is_known_ident

bvanjoi committed:
rm swc_ecma_lexer/TokenKind in lexer state update

bvanjoi committed:
rm swc_ecma_lexer/TokenType in lexer state update

bvanjoi commented:
The performance gains demonstrate a compelling case for implementation. 

Let me explain my idea:
- First, split `Token` into `TokenKind` (the type of token, which is still referred to as `Token` in this PR) and `TokenValue`. This means we can't retrieve the correct combination of `TokenKind` and `TokenValue` after parsing, because the `TokenValue` has already been consumed during parsing.
- And then, these changes mean there are now two different types of `Token` between `swc_ecma_lexer` and `swc_ecma_parser`: one contains the `TokenValue`, while the other does not.
- The points above mean that using a single Parser between `swc_ecma_lexer` and `swc_ecma_parser` is more complicated. We would need to create a trait to handle different types of `Token` and manage their behaviors, and the ROI for this seems low.
- So, I made the following changes: I copied the `swc_ecma_parser/parser` into `swc_ecma_lexer`, meaning that `swc_ecma_lexer` now includes the `Parser` itself. I also created a `swc_ecma_parser/lexer` folder, which contains a lexer that serves only the `swc_ecma_parser/Parser` and is not exposed externally. With these changes, if you want to generate the correct tokens, you should use `swc_ecma_lexer/{Lexer, Parser}`, while the `swc_ecma_parser/Parser` is now only for quickly generating the AST.
- Based on these changes, I’ve removed some tests that ensured the lexer worked correctly in `swc_ecma_parser` and switched to using `swc_ecma_lexer` in `swc_fast_ts_strip`.
- Another change is simply replacing `Token` with `TokenKind` and `TokenValue`.

@kdy1  Do you agree with these changes? If so, I will fix CI and make it ready for review.

CPunisher commented:
I have seen that the two lexers and two parsers in `swc_ecma_lexer` and `swc_ecma_parser`. Do we need maintain all of them or one pair is WIP while another is released?

bvanjoi commented:
> The performance gains demonstrate a compelling case for implementation.
> 
> Let me explain my idea:
> 
> * First, split `Token` into `TokenKind` (the type of token, which is still referred to as `Token` in this PR) and `TokenValue`. This means we can't retrieve the correct combination of `TokenKind` and `TokenValue` after parsing, because the `TokenValue` has already been consumed during parsing. (edit: Unless the `TokenValue` is stored elsewhere, but this is unnecessary for parsing.)
> * And then, these changes mean there are now two different types of `Token` between `swc_ecma_lexer` and `swc_ecma_parser`: one contains the `TokenValue`, while the other does not.
> * The points above mean that using a single Parser between `swc_ecma_lexer` and `swc_ecma_parser` is more complicated. We would need to create a trait to handle different types of `Token` and manage their behaviors, and the ROI for this seems low.
> * So, I made the following changes: I copied the `swc_ecma_parser/parser` into `swc_ecma_lexer`, meaning that `swc_ecma_lexer` now includes the `Parser` itself. I also created a `swc_ecma_parser/lexer` folder, which contains a lexer that serves only the `swc_ecma_parser/Parser` and is not exposed externally. With these changes, if you want to generate the correct tokens, you should use `swc_ecma_lexer/{Lexer, Parser}`, while the `swc_ecma_parser/Parser` is now only for quickly generating the AST.
> * Based on these changes, I’ve removed some tests that ensured the lexer worked correctly in `swc_ecma_parser` and switched to using `swc_ecma_lexer` in `swc_fast_ts_strip`.
> * Another change is simply replacing `Token` with `TokenKind` and `TokenValue`.
> 
> @kdy1 Do you agree with these changes? If so, I will fix CI and make it ready for review.

bvanjoi committed:
fix ci

bvanjoi commented:
> Do we need maintain all of them

Yes, I think so. Perhaps we can find a better way to integrate it.

kdy1 commented:
Why do we need two separate parser/lexer/token types? 
What's the problem you see in the token definition in https://github.com/swc-project/swc/pull/10397 ?

bvanjoi commented:
> Why do we need two separate **token** types?

The `value` field defined in `Token`([source](https://github.com/swc-project/swc/pull/10397/files#diff-b9ff6e1704d57c6d063326c9d5a20588f5b9816bbdccb6e8a05ec664c6f481beR22)) is not required for parsing. Moving it to the outer struct would reduce memory usage.

Additionally, the `had_line_break` field in `Token` during parsing also unnecessary and can be removed in the future.

> Why do we need two separate ** lexer** ?

Since the lexical analysis implementations differ, `swc_ecma_lexer/lexer` guarantees token correctness while `swc_ecma_parser/lexer` does not maintain the same level of validation.

> Why do we need two separate **parser**?

Using a single parser to handle different lexical logic introduces unnecessary development complexity. While I've separated them for now, I believe there should be a better way to consolidate this logic.

CPunisher commented:
> The `value` field defined in `Token`([source](https://github.com/swc-project/swc/pull/10397/files#diff-b9ff6e1704d57c6d063326c9d5a20588f5b9816bbdccb6e8a05ec664c6f481beR22)) is not required for parsing.

I'm not sure why `TokenValue` is not required for parsing. For example, doesn't an identifier token required for constructing identifier ast node?

CPunisher commented:
> * This means we can't retrieve the correct combination of `TokenKind` and `TokenValue` after parsing, because the `TokenValue` has already been consumed during parsing(edit: Unless the `TokenValue` is stored elsewhere, but this is unnecessary for parsing)

I guess this is the root cause of the splitting. But I'm unclear about it. Could you please provide more detailed examples?

bvanjoi commented:
> I'm not sure why TokenValue is not required for parsing. 

`TokenValue` is not required **in the `Token`**. For example:

```
a + b
^        --> `token: ident`, `lexer.state.token_value: a`
    ^    --> `token: ident`, `lexer.state.token_value: b`
    
// rather than
a + b
^        --> `token: token_kind: ident, token_value: a`
    ^    --> `token: token_kind: ident, token_value: a` // we dont store `token_value` in each `token`
```

kdy1 commented:
> The value field defined in Token([source](https://github.com/swc-project/swc/pull/10397/files#diff-b9ff6e1704d57c6d063326c9d5a20588f5b9816bbdccb6e8a05ec664c6f481beR22)) is not required for parsing. Moving it to the outer struct would reduce memory usage.
>
> Additionally, the had_line_break field in Token during parsing also unnecessary and can be removed in the future.

I understood, but I think the gain is too small if we have to maintain two tokens/lexers/parsers. My guess is that #10397 is the sweet point, but not sure.

bvanjoi commented:
>  In other words, what's the difference between them now

The `swc_ecma_parser/Parser` implementation prioritizes AST generation performance at the cost of potential token validation gaps, whereas `swc_ecma_lexer/Parser` enforces strict token-level correctness guarantees during AST construction.

> I think the gain is too small if we have to maintain two tokens/lexers/parsers

I suggest considering landing this change for three reasons:
- ​Using a separate lexer and parser between `swc_ecma_parser` and `swc_ecma_lexer` is the foundation for future optimizations (such as removing `TokenContexts`, which requires using a custom lexer). I think with this change we can eventually achieve the same performance as `oxc/parser`.
​- Strict token validation is not commonly needed, and sacrificing performance improvements for it may not be a good trade-off.
- ​We can reuse code by creating a shared crate (e.g., `swc_ecma_lexer_common`) to store common functions and traits.

bvanjoi commented:
> > In other words, what's the difference between them now
> 
> The `swc_ecma_parser/Parser` implementation prioritizes AST generation performance at the cost of potential token validation gaps, whereas `swc_ecma_lexer/Parser` enforces strict token-level correctness guarantees during AST construction.
> 
> > I think the gain is too small if we have to maintain two tokens/lexers/parsers
> 
> I suggest considering landing this change for three reasons:
> 
> * ​Using a separate lexer and parser between `swc_ecma_parser` and `swc_ecma_lexer` is the foundation for future optimizations (such as removing `TokenContexts`, which requires using a custom lexer; and remove the `had_line_break ` field in `Token`, which need different `Token` type). I think with this change we can eventually achieve the same performance as `oxc/parser`.
> * Strict token validation is not commonly needed, and sacrificing performance improvements for it may not be a good trade-off.
> * ​We can reuse code by creating a shared crate (e.g., `swc_ecma_lexer_common`) to store common functions and traits.

bvanjoi commented:
> I'm not too clear about Strict token validation. Generating correct AST is always the foundation for parser.

Consider the tokenization of `<` in `<div/>` when parsing a JSX file. Under strict parsing rules, this would be classified as a `Token::JsxTagStart`. However, in practice, treating it as a simple `Token::LessThan` is sufficient for generating a correct AST. The stricter approach requiring `Token::JsxTagStart` forces us to maintain unnecessary contextual information purely for AST construction purposes.

> why we still keep `swc_ecma_lexer/Parser` in the project

The primary purpose of `swc_ecma_lexer/Parser` is to ensure the lexer can function reliably, particularly for cases involving `try_parse` methods that may require multiple lexing attempts. Unlike a single-pass lexer, this implementation needs to handle scenarios where tokens cannot be fully determined in one pass.

(Alternatively, we might consider eventually removing `swc_ecma_lexer/Parser` entirely and simply exposing a `lexical_analysis` function that handles tokenization independently.)

> Do you mean copy the new Lexer and Parser, optimize on top of the new ones, and replace the original ones once they are completed

I'm having difficulty understanding the intended meaning here. Could you please clarify?

kdy1 commented:
Ok, I got the idea. Would it be possible to use a single struct with const generic or generic to make maintenance easy? I'm worried about the maintenance burden.

CPunisher commented:
> Ok, I got the idea and I like it. Would it be possible to use a single struct with const generic or generic to make maintenance easy? I'm worried about the maintenance burden.

Agree. I expect finally we should eliminate all duplication and get a clear lexer and a clear parser (just like now, and faster), but we still make the code in the process as maintainable as possible.

kdy1 commented:
If it's inevitable at the moment, I think we can merge this PR as is and refactor these with other PRs, but I think we should at least have concrete plan for future deduplication.

CPunisher commented:
Currently most of the code in these two set of lexers and parsers are identical, which is exactly not elegant...

bvanjoi commented:
> I'm worried about the maintenance burden.

I am also worried about the maintenance cost: At present, it seems troublesome to abstract `swc_ecma_lexer/Parser` and `swc_ecma_parser/Parser` into an independent trait. My suggestion is:
- Create a library `swc_ecma_lexer_common`, and then put the common parts of `swc_ecma_lexer` and `swc_ecma_parser` into it, and `swc_ecma_parser` will no longer depend on `swc_ecma_lexer`.
- Gradually sort out the reusable logic of the two and implement it in `swc_ecma_lexer_common`.

What do you think?

kdy1 commented:
It sounds acceptable for now, but I want to hear opinion from @CPunisher, too.

CPunisher commented:
> > I'm worried about the maintenance burden.
> 
> I am also worried about the maintenance cost: At present, it seems troublesome to abstract `swc_ecma_lexer/Parser` and `swc_ecma_parser/Parser` into an independent trait. My suggestion is:
> 
> * Create a library `swc_ecma_lexer_common`, and then put the common parts of `swc_ecma_lexer` and `swc_ecma_parser` into it, and `swc_ecma_parser` will no longer depend on `swc_ecma_lexer`.
> * Gradually sort out the reusable logic of the two and implement it in `swc_ecma_lexer_common`.
> 
> What do you think?

It's a not bad way. But I **personally** prefer merging everything back into one crate and control the public items (for example, export `swc_ecma_lexer/{Lexer, Parser}` and `swc_ecma_parser/Parser`). In my opinion, this makes code reuse easier, narrow down the scope of code changes and avoid spreading things across the whole project. I'm not sure if this is a more reasonable way though.

bvanjoi commented:
At this stage, I don’t have a strong preference (though I’ll provide suggestions if technical advantages arise). Do you think we could use `swc_ecma_lexer/src/common` to store these shared structures?

kdy1 commented:
So... @CPunisher, do you think this PR is good to merge in its current state, or should we wait for the `swc_ecma_lexer/src/common` refactoring?

CPunisher commented:
> So... @CPunisher, do you think this PR is good to merge in its current state, or should we wait for the `swc_ecma_lexer/src/common` refactoring?

Considering the improvement vs the code complexity, I think it's better to refactor first.

bvanjoi committed:
chore(es/parser): common lexer::comment_buffer

bvanjoi committed:
chore(es/parser): common lexer::whiltespace

bvanjoi committed:
chore(es/parser): common lexer::LexResult

bvanjoi committed:
chore: common lexer::{Char, CharIter, CharExt}

bvanjoi committed:
chore: common syntax

bvanjoi committed:
chore: common Context

bvanjoi committed:
chore: common input::Tokens

bvanjoi committed:
chore: rm useless input

bvanjoi commented:
I plan to refactor both the `Lexer` and `Parser` into trait-based implementations, and will remove the draft status once completed.

bvanjoi committed:
chore(es/parser): common lexer:State

bvanjoi committed:
inline methods in TokenKind

bvanjoi committed:
chore(es/lexer): common utils of lexer, part1

bvanjoi committed:
chore(es/lexer): common emit_error of lexer

bvanjoi committed:
chore(es/lexer): common `skip_line_comment`

bvanjoi committed:
chore(es/lexer): common `skip_space`

bvanjoi committed:
chore(es/lexer): rm `store_comment`

bvanjoi committed:
chore(es/lexer): common number, part1

bvanjoi committed:
chore(es/lexer): common number, part2

bvanjoi committed:
chore(es/lexer): common `consume_pending_comments`

bvanjoi committed:
chore(es/lexer): common `read_jsx_word`

bvanjoi committed:
chore(es/lexer): common `read_jsx_str`

bvanjoi committed:
chore(es/lexer): common read_token, part1

bvanjoi committed:
chore(es/lexer): common read_token, part2

bvanjoi committed:
chore(es/lexer): common read_token, part3

bvanjoi committed:
chore(es/parser): common `ExprExt`

bvanjoi committed:
chore(es/parser): delete test in lexer/parser

bvanjoi committed:
chore(es/parser): common `with_state`

bvanjoi committed:
chore(es/parser): common buffer

bvanjoi committed:
chore(es/parser): common `WithCtx`

bvanjoi committed:
chore(es/lexer): some traits

bvanjoi committed:
chore(es/parser): common `emit_error`

bvanjoi committed:
chore(es/parser): common helpers

bvanjoi committed:
chore(es/parser): common `Verifier`

bvanjoi committed:
chore(es/parser): common `verify_expr`

bvanjoi committed:
chore(es/parser): common `parse_lit`

bvanjoi committed:
chore(es/parser): common `parse_ident_name`

bvanjoi committed:
chore(es/parser): common `make_decl_declare`

bvanjoi committed:
chore(es/parser): common parse private name

bvanjoi committed:
chore(es/parser): common parse_ident

bvanjoi committed:
chore(es/parser): common `is_start_of_expr`

bvanjoi committed:
chore(es/parser): binding ident and ts modifier

bvanjoi committed:
chore(es/parser): common `parse_ts_this_type_node`

bvanjoi committed:
chore(es/parser): common `parse_tpl_element`

bvanjoi committed:
chore(es/parser): parse jsx name

bvanjoi committed:
chore(es/parser): parse jsx empty expr

bvanjoi committed:
chore(es/parser): common `eat_any_ts_modifier`

bvanjoi committed:
chore(es/parser): common `parse_jsx_text`

bvanjoi committed:
chore(es/parser): common reparse_expr_as_pat

bvanjoi committed:
chore(es/parser): common parse_paren_items_as_params

bvanjoi commented:
I've identified minor performance regressions in both the `lexer` and `fast-strip`. Based on preliminary analysis, these appear to be related to trait code generation. However, I'm choosing to disregard these regressions for two reasons: (1) the impact is negligible (nanosecond-level differences), and (2) the variations may simply reflect differences between Rust compiler versions rather than actual performance issues

bvanjoi committed:
chore(es/parser): common parse_assignment_expr

bvanjoi committed:
chore(es/parser): common parse_ts_entity_name

bvanjoi committed:
chore(es/parser): common parse_ty, part1

kdy1 commented:
I think small regression of them is acceptable at this time. I can optimizer further at the future.

bvanjoi committed:
chore(es/parser): common parse_ty, part2

bvanjoi committed:
chore(es/parser): common parse_ty, part3

bvanjoi committed:
chore(es/parser): use assoc const item for token

bvanjoi committed:
chore(es/parser): common parse_ty, part4

bvanjoi committed:
fix rebase issue

bvanjoi committed:
chore(es/parser): common parse_ty, part5

bvanjoi committed:
chore(es/parser): common parse jsx

bvanjoi committed:
chore(es/parser): common parse expr, part1

bvanjoi committed:
chore(es/parser): common parse stmt, part1

bvanjoi committed:
chore(es/parser): common parse_ident

bvanjoi committed:
chore(es/parser): common parse pattern, part2

bvanjoi committed:
chore(es/parser): common parse_ty, part6

bvanjoi committed:
chore(es/parser): common parse pattern, part3

bvanjoi committed:
chore(es/parser): common parse_ty, part7

bvanjoi committed:
chore(es/parser): common parse_ty, part8

bvanjoi committed:
chore\(es/parser\): common parse stmt, part2

bvanjoi committed:
chore(es/parser): common parse expr

bvanjoi committed:
chore(es/parser): common parse stmt

bvanjoi committed:
chore(es/parser): common parse class and fn

bvanjoi commented:
Completed, marked as ready for review.

bvanjoi commented:
While all tests have passed, this significant change may introduce some unintended side effects. However, we can address issues that arise on a case-by-case basis.

bvanjoi commented:
@kdy1  It seems Copilot isn't providing helpful review comments. Do you have any suggestions? Alternatively, can we proceed with this change and I can make further improvements base this?

kdy1 commented:
You can ignore Copilot. I wanted to see if it would generate a good review

kdy1 committed:
Create twenty-rocks-deny.md

kdy1 commented on crates/swc_ecma_parser/src/lexer/util.rs:
```diff
@@ -0,0 +1,119 @@
+                        self.state.had_line_break = true;
+                    }
+                    // Skip multibyte characters
+                    pos += c.len_utf8() - 1; // -1은 아래 증가분 고려
```
Can you remove Korean comments?

bvanjoi committed:
refactor(es/lexer): delete Korean comment

kdy1 commented on crates/swc_ecma_parser/src/lexer/util.rs:
```diff
@@ -0,0 +1,119 @@
+
+            if was_star && b == b'/' {
+                // Found comment end: "*/"
+                self.input.bump_bytes(pos + 1); // 종료 '/' 포함해서 이동
```
Please delete korean comments

kdy1 commented on crates/swc_ecma_parser/src/lexer/token.rs:
```diff
@@ -0,0 +1,1493 @@
+    JSXText,
+    // Identifiers and keyword
+    Ident,
+    // Reserved keyword tokens (starting from 100)
```
Does this start from 100?

kdy1 commented on crates/swc_ecma_parser/src/lexer/token.rs:
```diff
@@ -0,0 +1,1493 @@
+    Yield,
+    Module,
+
+    // TypeScript-related keywords (starting from 150)
```
Same Q here

kdy1 commented on crates/swc_ecma_parser/src/lexer/token.rs:
```diff
@@ -0,0 +1,1493 @@
+        t >= Token::Abstract as u8 && t <= Token::Target as u8
+    }
+
+    pub(crate) const fn is_word(&self) -> bool {
```
Is there a reason you used `&self `instead of `self` here?

kdy1 commented on crates/swc_ecma_lexer/src/common/lexer/token.rs:
```diff
@@ -0,0 +1,569 @@
+    fn take_known_ident(&self) -> Atom;
+
+    fn starts_expr(&self) -> bool;
+    fn to_string(&self, buffer: &Self::Buffer) -> String;
```
This does not seem to be used

bvanjoi replied:
see: https://github.com/swc-project/swc/pull/10399/files#diff-d22806d12c2473292860573463bef072dd0c36c2612083e03e455645de5c4b76R132

kdy1 commented on crates/swc_ecma_parser/src/lexer/token.rs:
```diff
@@ -0,0 +1,1493 @@
+        }
+    }
+
+    pub(crate) fn to_string(self, value: Option<&TokenValue>) -> String {
```
Can you add `#[cold]`?

kdy1 commented on crates/swc_ecma_parser/src/lexer/token.rs:
```diff
@@ -0,0 +1,1493 @@
+            || (*self).is_keyword()
+    }
+
+    pub(crate) fn as_word_atom(&self, value: Option<&TokenValue>) -> Option<Atom> {
```
Same here, `Token` is `: Copy`. You don't need `&self` and needless referencing of Copy types end to make things slower

kdy1 commented on crates/swc_ecma_parser/src/lexer/token.rs:
```diff
@@ -0,0 +1,1493 @@
+            || (t >= Token::Plus as u8 && t <= Token::Ampersand as u8)
+    }
+
+    pub(crate) fn as_bin_op(&self) -> Option<swc_ecma_ast::BinaryOp> {
```
Is `&self` intentional?

bvanjoi committed:
refactor(ecma/lexer): self by value

CPunisher commented:
I think current structure is mostly acceptable. I have some other findings/questions that we can improve in the future.
1. Most parser logic in `swc_ecma_parser` have been moved to `swc_ecma_lexer::common` and there are only some trivial parser logic and tests now. We may separate the tests to other folders.
2. The `Parser` in `swc_ecma_lexer` now is only used by `swc_fast_ts_strip`. Does it mean if we refactor `swc_fast_ts_strip` to use the `Parser` in `swc_ecma_parser`, then we can only keep one `Parser` in the project?

bvanjoi commented:
> then we can only keep one Parser in the project

I don't think so, because generating the correct tokens requires `try_parse` logic, which means we can't get it right in a single pass and need a parser to attempt multiple times.

kdy1 commented:
Sorry for the delay, I'm still reviewing

kdy1 commented on crates/swc_ecma_parser/src/lexer/mod.rs:
```diff
@@ -0,0 +1,348 @@
+        std::mem::transmute::<&str, &'a str>(s)
+    }
+
+    fn input_uncons_while(&mut self, f: impl FnMut(char) -> bool) -> &'a str {
```
Can't we bound `&mut self` instead? This feels like an unnecessary transmute

bvanjoi replied:
The latest commit resolve this issue, though it introduces changes.

kdy1 commented on crates/swc_ecma_parser/src/lexer/mod.rs:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    #[inline(always)]
+    unsafe fn input_slice(&mut self, start: BytePos, end: BytePos) -> &'a str {
```
Also here

kdy1 reviewed (dismissed):
Let's improve iteratively. Thank you so much for your hard work!

kdy1 commented:
Can you rebase?

bvanjoi committed:
refactor(es/lexer): add lifetime for Input tarit

kdy1 committed:
Create kind-weeks-study.md
